lib/src/controller/fitbit_screen_controller.dart

```diff
@@ -0,0 +1,163 @@
+        connectFitbit: _connectFitbit);
+  }
+
+  void _connectFitbit() {
+    //TODO(): Add fetch from secure db & check for token validity
+    if (_accessToken == '') {
+      var authenticationService = getIt<AuthenticationService>();
+      authenticationService.fitbitSignIn().then((value) {
+          if (value != null) {
+            _userId = value.userID;
+            _accessToken = value.fitbitAccessToken;
+            _refreshToken = value.fitbitRefreshToken;
+            AuthUtils.saveFitbitUserToDB(_accessToken, _refreshToken, _userId);
+            developer.log('Fitbit Credentials saved');
+          } else {
+            developer.log('No FitBit Credentials recieved');
+          }
+      });
+    } 
+    _fetchData();
+  }
+
+  void _fetchData() {
+    var date = DateTime.now();
+    //TODO(): Add Intraday data
+    var daysSinceSync = _lastSync.difference(date).inDays;
+    if (daysSinceSync > 0) {
+      var clientId = AppConfig.shared.currentConfig.fitbitClientId;
+      var clientSecret = AppConfig.shared.currentConfig.fitbitClientSecret;
+      var credentials = FitbitCredentials(userID: _userId,
+                                          fitbitAccessToken: _accessToken, fitbitRefreshToken: _refreshToken);
+
+      // Activity Data
+      for (int day in Iterable.generate(daysSinceSync)) {
+        FitbitActivityDataManager fitbitActivityDataManager = FitbitActivityDataManager(
+              clientID: clientId, clientSecret: clientSecret);
+        FitbitActivityAPIURL fitbitActivityAPIURL = FitbitActivityAPIURL.day(
+                                          date: _lastSync.add(Duration(days: day)),
+                                          fitbitCredentials: credentials,
+                                        );
+        _fetchAndConvertData(fitbitActivityDataManager, fitbitActivityAPIURL);
+      }
+
+      // Activity Timeseries Data
+      FitbitActivityTimeseriesDataManager fitbitActivityTSDataManager = FitbitActivityTimeseriesDataManager(
+            clientID: clientId, clientSecret: clientSecret);
+      for (Resource r in Resource.values) {
+        FitbitActivityTimeseriesAPIURL fitbitActivityTSAPIURL = FitbitActivityTimeseriesAPIURL.dateRangeWithResource(
+                                        startDate: _lastSync,
+                                        endDate: DateTime.now(),
+                                        fitbitCredentials: credentials,
+                                        resource: r
+                                      );
+        _fetchAndConvertData(fitbitActivityTSDataManager, fitbitActivityTSAPIURL);
+      }
+      
+      // Breathing Rate Data
+      FitbitBreathingRateDataManager fitbitBRDataManager = FitbitBreathingRateDataManager(
+            clientID: clientId, clientSecret: clientSecret);
+      FitbitBreathingRateAPIURL fitbitBRAPIURL = FitbitBreathingRateAPIURL.dateRange(
+                                        startDate: _lastSync, endDate: DateTime.now(),
```

Comment:
Should we move all these fetchAndConvertData blocks into their individual functions and call them in parallel?

Reply:
Yes, good idea. I'll update in the next CL